README.md
# Arcade Game Project

[![ArcadeGame last commit](https://img.shields.io/github/last-commit/google/skia.svg)](https://github.com/jdmedlock/arcadegame)
<br/>
[![Packagist](https://img.shields.io/packagist/l/doctrine/orm.svg)](https://github.com/jdmedlock/arcadegame/)

## Table of Contents

* [Overview](#overview)
* [How to Play](#how-to-play)
* [Player UI Feature](#player-ui-features)
* [Change Log](#change-log)
* [Contributing](#contributing)
* [Authors](#authors)
* [License](#license)

## Overview

The Arcade Game project was created as part of the Exploring JS section of the
[Udacity Front-End Web Developer Nanodegree Program](https://www.udacity.com/course/front-end-web-developer-nanodegree--nd001). The
purpose of this assignment is to demonstrate mastery of the Javascript classes
and objects, but also to build on creating responsive and accessible web apps.

You can play the game here --> [Arcade Game](https://jdmedlock.github.io/arcadegame/)

## How to Play

This game is a browser-based game similar to the classic
[Frogger](https://en.wikipedia.org/wiki/Frogger) arcade game developed by
Konami in 1981.

In our game the objective is to move the players avatar from the starting point
at the bottom of the game board to the water at the top of the board - without
getting hit by a bug.

The bugs move at varying speeds from left-to-right across the paved lanes on
the board. The player uses keyboard arrow keys to move the avatar left, right,
up and down to move one block at a time.

When the game starts bugs always are always in the first column in their
respective row. However, the player avatar always starts in the bottom row, but
at a random column position for each game.

The game is won when the player's avatar reaches the water. However, if the
avatar collides with a bug the game is lost. At the end of a game they players
score is updated and the avatar is moved to the starting square.

## Player UI Features

In addition to the basic game play several UI components have been implemented
to provide the player with features to improve the overall experience.

* Score - The players score is represented as 'www / ggg' where 'www' is the
number of games won and 'ggg' is the number of games played

* Gems - These are collectibles that are added and removed based on the
players win percentage. A gem is added whenever the players win percentage
advances into the next higher percentile group. A gem is taken away when the win
percentage drops into the next lower percentile groups. There are four percentile
groups each representing a score as shown below

| Gems Accumulated | Win Percentile |
|:----------------:|:--------------:|
| 0 | 0-9% |
| 1 | 10-32% |
| 2 | 33-65% |
| 3 | 66-100% |

## Building & Running the Game

### Building on a Local Computer

To build the app on your local computer open a new terminal window and follow
the steps shown below.

| # | Command | Comments |
|:-:|:--------|:---------|
| 1 | `cd <parent-destination-directory>` | Navigate to the directory that will contain the application's sub-directory. |
| 2 | `git clone https://github.com/jdmedlock/arcadegame.git` | This creates the application sub-directory and populates it with the source code and documentation for the current release maintained on GitHub. |

At this point the app will be located at the directory
`<parent-destination-directory>/arcadegame` on your local computer.

This app has no external dependencies to any libraries or modules other than
the Javascript files in the `js` directory.

### Running on a Local Computer

When you have successfully downloaded the app from GitHub it can be run from
your browsers File menu by opening `index.html` located in the application
directory you created above.

## Change Log

For more information see [Change Log](https://github.com/jdmedlock/arcadegame/blob/development/CHANGELOG.md)

## Contributing

See [Contributing](https://github.com/jdmedlock/arcadegame/blob/development/CONTRIBUTING.md)
and our [Collaborator Guide](https://github.com/jdmedlock/arcadegame/blob/development/COLLABORATOR_GUIDE.md).

## Authors

Developers on this project can be found on the [Contributors](https://github.com/jdmedlock/arcadegame/graphs/contributors) page of this repo.

## License

[MIT](https://tldrlegal.com/license/mit-license)



CHANGELOG.md
# Change Log

### 1.0.0 - July 14, 2018

- Initial release

### 1.0.1 - July 14, 2018

- Bug fixes and corrections from code review

  - #19 Incorrect use of `player.handleInput(...)` within Player class
  - #20 Player avatar can be moved off screen
  - #21 Player-Enemy collisions are not being detected soon enough
  - #22 Add local build/load/run instructions for a local machine to README.md
  - #23 Remove unused `update()` function definition from Player.js


js/Enemy.js
/**
 * @description Enemy defines an enemy sprite and controls its movement on the
 * game board. The game engine requires it to contain `render` and `update`
 * methods.
 * @class Enemy
 */
class Enemy {

  /**
   * @description Creates an instance of the Enemy class and establishes the
   * required object variables.
   * @param {Number} row Number of the game board row this enemy is constrained
   * to.
   * @memberof Enemy
   */
  constructor(row) {
    // The image/sprite for our enemies, this uses
    // a helper we've provided to easily load images
    this.sprite = 'images/enemy-bug.png';

    this.rowConstraint = row;
    this.x = MIN_ROW_BOUNDARY;
    this.y = (this.rowConstraint * CELL_HEIGHT) - 20;
    this.startRow = 1;
    this.endRow = 3;
  }

  /**
   * @description Retrieve the current position of the enemy avatar
   * @returns {Object} An object containing the x and y coordinates of the
   * player avatar. 
   * @memberof Enemy
   */
  getPosition() {
    return {
      x: this.x,
      y: this.y
    };
  }

  /**
   * @description Update the enemy's position. This method is required by the
   * game engine.
   * @param {Number} dt a time delta between ticks of the game clock
   * @memberof Enemy
   */
  update(dt) {
    // You should multiply any movement by the dt parameter
    // which will ensure the game runs at the same speed for
    // all computers.
    this.x = this.x >= MAX_ROW_BOUNDARY ? MIN_ROW_BOUNDARY : this.x;
    this.x = this.x + ((50 * dt) * this.rowConstraint);
  }

  /**
   * @description Draw this enemy on the screen. This method is required by the
   * game engine.
   * @memberof Enemy
   */
  render() {
    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
  }
}

js/Player.js

/**
 * @description Player defines the player sprite and controls its movement
 * game board
 * @class Player
 */
class Player {
  /**
   * @description Creates an instance of Player and creates and initializes
   * object variables. The game engine requires it to contain `render` and
   * `update` methods.
   * @memberof Player
   */
  constructor() {
    // Variables applied to each of our instances go here,
    // we've provided one for you to get started

    // The image/sprite for our player, this uses
    // a helper we've provided to easily load images
    this.sprite = 'images/char-boy.png';
    this.startingRow = MAX_ROWS - 1;
    this.resetPosition();

    // This listens for key presses and sends the keys to your
    // Player.handleInput() method. You don't need to modify this.
    document.addEventListener('keyup', (event) => {
      var allowedKeys = {
        27: 'esc',
        37: 'left',
        38: 'up',
        39: 'right',
        40: 'down'
      };

      this.handleInput(allowedKeys[event.keyCode]);
    });
  }

  /**
   * @description Reset the player avatar to its starting position on the
   * game board
   * @memberof Player
   */
  resetPosition() {
    this.startingCol = this.getRandomInt(1, MAX_COLS);
    this.x = (this.startingCol - 1) * CELL_WIDTH;
    this.y = (this.startingRow * CELL_HEIGHT) - PLAYER_YOFFSET;
  }

  /**
   * @description Retrieve the current position of the player avatar
   * @returns {Object} An object containing the x and y coordinates of the
   * player avatar.
   * @memberof Player
   */
  getPosition() {
    return {
      x: this.x,
      y: this.y
    };
  }

  /**
   * @description Calculate a random number between specified minimum and
   * maximum integers, inclusive. This function was copied from the example at
   * [MDN Javascript Math builtin object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random)
   * @param {Number} max
   * @returns {Number} A number in the range 0-`max`
   */
  getRandomInt(min, max) {
    min = Math.ceil(min);
    max = Math.floor(max);
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }

  /**
   * @description Convert a y coordinate value in the canvas to a row number in
   * the game grid.
   * @param {Number} y Value of the y coordinate in the canvas
   * @returns {Number} Row number within the game grid
   * @memberof Player
   */
  convertToRowNo(y) {
    return Math.ceil(this.y / CELL_HEIGHT);
  }

  /**
   * @description Draw the player on the screen. This method is required by the
   * game engine.
   * @memberof Player
   */
  render() {
    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
  }

  /**
   * @description Process a request to move the player avatar one cell on the
   * game board. Attempts to move past the left, right, top, or bottom
   * boundaries are ignored. The x and y coordinates of the players avatar
   * are calculated here rather than in the update() function to eliminate the
   * dependency on a saved direction state in update().
   * @param {String} keyCode Direction to move the player avatar. Acceptable
   * values are `left`, `right`, `up`, or `down`.
   * @memberof Player
   */
  handleInput(keyCode) {
    const continueButton = event.target.closest('#continue-button');
    if (continueButton === null) {
      switch (keyCode) {
        case 'left':
          this.x = this.x <= MIN_ROW_BOUNDARY ? this.x : this.x - CELL_WIDTH;
          break;
        case 'up':
          this.y = this.y <= MIN_COL_BOUNDARY ? this.y : this.y - CELL_HEIGHT;
          break;
        case 'right':
          this.x = this.x >= MAX_ROW_BOUNDARY ? this.x : this.x + CELL_WIDTH;
          break;
        case 'down':
          this.y = this.y >= MAX_COL_BOUNDARY ? this.y : this.y + CELL_HEIGHT;
          break;
        default:
          break;
      }
    }
  }

  /**
   * @description Check to see if the player has won the current game
   * @returns {Boolean} `true` if the game has been won
   * @memberof Player
   */
  hasWon() {
    if (this.convertToRowNo(this.y) === WIN_ROW) {
      this.resetPosition();
      return true;
    }
    return false;
  }
}

js/engine.js
/* Engine.js
 * This file provides the game loop functionality (update entities and render),
 * draws the initial game board on the screen, and then calls the update and
 * render methods on your player and enemy objects (defined in your app.js).
 *
 * A game engine works by drawing the entire game screen over and over, kind of
 * like a flipbook you may have created as a kid. When your player moves across
 * the screen, it may look like just that image/character is moving or being
 * drawn but that is not the case. What's really happening is the entire "scene"
 * is being drawn over and over, presenting the illusion of animation.
 *
 * This engine makes the canvas' context (ctx) object globally available to make
 * writing app.js a little simpler to work with.
 */

var Engine = (function (global) {
	/* Predefine the variables we'll be using within this scope,
	 * create the canvas element, grab the 2D context for that canvas
	 * set the canvas elements height/width and add it to the DOM.
	 */
	const doc = global.document;
	const	win = global.window;
	const	canvas = doc.createElement('canvas');
	const	ctx = canvas.getContext('2d');
	let	lastTime;
	let	shouldStop = false;

	canvas.width = 505;
	canvas.height = 606;
	doc.getElementById('game-board').appendChild(canvas);

	/* This function serves as the kickoff point for the game loop itself
	 * and handles properly calling the update and render methods.
	 */
	function main() {
		/* Get our time delta information which is required if your game
		 * requires smooth animation. Because everyone's computer processes
		 * instructions at different speeds we need a constant value that
		 * would be the same for everyone (regardless of how fast their
		 * computer is) - hurray time!
		 */
		var now = Date.now(),
			dt = (now - lastTime) / 1000.0;

		/* Call our update/render functions, pass along the time delta to
		 * our update function since it may be used for smooth animation.
		 */
		update(dt);
		render();

		/* Set our lastTime variable which is used to determine the time delta
		 * for the next time this function is called.
		 */
		lastTime = now;

		/* Use the browser's requestAnimationFrame function to call this
		 * function again as soon as the browser is able to draw another frame.
		 */
		if (shouldStop) {
			reset();
			shouldStop = false;
		}
		if (!shouldStop) {
			win.requestAnimationFrame(main);
		}
	}

	/* This function does some initial setup that should only occur once,
	 * particularly setting the lastTime variable that is required for the
	 * game loop.
	 */
	function init() {
		reset();
		lastTime = Date.now();
		main();
	}

	/* This function is called by main (our game loop) and itself calls all
	 * of the functions which may need to update entity's data. Based on how
	 * you implement your collision detection (when two entities occupy the
	 * same space, for instance when your character should die), you may find
	 * the need to add an additional function call here. For now, we've left
	 * it commented out - you may or may not want to implement this
	 * functionality this way (you could just implement collision detection
	 * on the entities themselves within your app.js file).
	 */
	function update(dt) {
		updateEntities(dt);
		shouldStop = checkCollisions();
		if (!shouldStop && player.hasWon()) {
			shouldStop = true;
			doc.getElementById('wins').innerText = game.incrementWins();
			game.openModal('You have won another game!');
		}
	}

	/**
	 * @description Check the position of the player avatar against that of
	 * all enemy avatars to see if any collisions have occurred. Collisions may
	 * occur when the enemies right side intersects with the left side of the
	 * player or if the right side of the player intersects with the left side
	 * of the enemy.
	 */
	function checkCollisions() {
		const playerPosition = player.getPosition();
		const playerXRight = Math.round(playerPosition.x) + CELL_WIDTH;
		for (let i = 0; i < allEnemies.length; i += 1) {
			const enemyPosition = allEnemies[i].getPosition();
			const enemyXRight = Math.round(enemyPosition.x) + CELL_WIDTH;

			// For Enemy-Player collision the following conditions must be met:
			// - (EnemyX + cell width) >= PlayerX
			// - (EnemyX + cell width) <= (PlayerX + 5)
			// - EnemyY = (PlayerY - icon offset)
			const isEnemyPlayerCollision =
				(enemyXRight >= Math.round(playerPosition.x)) &&
				(enemyXRight <= Math.round(playerPosition.x + 5)) &&
				(Math.round(enemyPosition.y) === Math.round(playerPosition.y - PLAYER_YOFFSET));

			// For player-enemy collision the following conditions must be met:
			// - (PlayerX + cell width) >= EnemyX
			// - (PlayerX + cell width) <= (EnemyX + 5)
			// - (PlayerY - icon offset) = EnemyY
			const isPlayerEnemyCollision =
				(playerXRight >= Math.round(enemyPosition.x)) &&
				(playerXRight <= Math.round(enemyPosition.x + 5)) &&
				(Math.round(playerPosition.y - PLAYER_YOFFSET) === Math.round(enemyPosition.y));

			if (isEnemyPlayerCollision || isPlayerEnemyCollision) {
				game.openModal('You have been defeated! So sad...');
				return true;
			}
		}
		return false;
	}

	/* This is called by the update function and loops through all of the
	 * objects within your allEnemies array as defined in app.js and calls
	 * their update() methods. It will then call the update function for your
	 * player object. These update methods should focus purely on updating
	 * the data/properties related to the object. Do your drawing in your
	 * render methods.
	 */
	function updateEntities(dt) {
		allEnemies.forEach(function (enemy) {
			enemy.update(dt);
		});
	}

	/* This function initially draws the "game level", it will then call
	 * the renderEntities function. Remember, this function is called every
	 * game tick (or loop of the game engine) because that's how games work -
	 * they are flipbooks creating the illusion of animation but in reality
	 * they are just drawing the entire screen over and over.
	 */
	function render() {
		/* This array holds the relative URL to the image used
		 * for that particular row of the game level.
		 */
		var rowImages = [
				'images/water-block.png', // Top row is water
				'images/stone-block.png', // Row 1 of 3 of stone
				'images/stone-block.png', // Row 2 of 3 of stone
				'images/stone-block.png', // Row 3 of 3 of stone
				'images/grass-block.png', // Row 1 of 2 of grass
				'images/grass-block.png' // Row 2 of 2 of grass
			],
			numRows = MAX_ROWS,
			numCols = MAX_COLS,
			row, col;

		// Before drawing, clear existing canvas
		ctx.clearRect(0, 0, canvas.width, canvas.height);

		/* Loop through the number of rows and columns we've defined above
		 * and, using the rowImages array, draw the correct image for that
		 * portion of the "grid"
		 */
		for (row = 0; row < numRows; row++) {
			for (col = 0; col < numCols; col++) {
				/* The drawImage function of the canvas' context element
				 * requires 3 parameters: the image to draw, the x coordinate
				 * to start drawing and the y coordinate to start drawing.
				 * We're using our Resources helpers to refer to our images
				 * so that we get the benefits of caching these images, since
				 * we're using them over and over.
				 */
				ctx.drawImage(Resources.get(rowImages[row]), col * CELL_WIDTH, row * CELL_HEIGHT);
			}
		}

		renderEntities();
	}

	/* This function is called by the render function and is called on each game
	 * tick. Its purpose is to then call the render functions you have defined
	 * on your enemy and player entities within app.js
	 */
	function renderEntities() {
		/* Loop through all of the objects within the allEnemies array and call
		 * the render function you have defined.
		 */
		allEnemies.forEach(function (enemy) {
			enemy.render();
		});

		player.render();
	}

	/* This function does nothing but it could have been a good place to
	 * handle game reset states - maybe a new game menu or a game over screen
	 * those sorts of things. It's only called once by the init() method.
	 */
	function reset() {
		doc.getElementById('plays').innerText = game.incrementPlays();
		player.resetPosition();
	}

	/* Go ahead and load all of the images we know we're going to need to
	 * draw our game level. Then set init as the callback method, so that when
	 * all of these images are properly loaded our game will start.
	 */
	Resources.load([
		'images/char-boy.png',
		'images/char-cat-girl.png',
		'images/char-horn-girl.png',
		'images/char-pink-girl.png',
		'images/char-princess-girl.png',
		'images/enemy-bug.png',
		'images/Gem Blue.png',
		'images/Gem Green.png',
		'images/Gem Orange.png',
		'images/grass-block.png',
		'images/Heart.png',
		'images/Key.png',
		'images/Rock.png',
		'images/Selector.png',
		'images/Star.png',
		'images/stone-block.png',
		'images/water-block.png',
	]);
	Resources.onReady(init);

	/* Assign the canvas' context object to the global variable (the window
	 * object when run in a browser) so that developers can use it more easily
	 * from within their app.js files.
	 */
	global.ctx = ctx;
})(this);

js/player.js

/**
 * @description Player defines the player sprite and controls its movement
 * game board
 * @class Player
 */
class Player {
  /**
   * @description Creates an instance of Player and creates and initializes
   * object variables. The game engine requires it to contain `render` and
   * `update` methods.
   * @memberof Player
   */
  constructor() {
    // Variables applied to each of our instances go here,
    // we've provided one for you to get started

    // The image/sprite for our enemies, this uses
    // a helper we've provided to easily load images
    this.sprite = 'images/char-boy.png';
    this.startingRow = MAX_ROWS - 1;
    this.resetPosition();
  }

  /**
   * @description Reset the player avatar to its starting position on the
   * game board
   * @memberof Player
   */
  resetPosition() {
    this.startingCol = this.getRandomInt(MAX_ROWS);
    this.x = this.startingCol * CELL_WIDTH;
    this.y = (this.startingRow * CELL_HEIGHT) - 10;
  }

  /**
   * @description Retrieve the current position of the player avatar
   * @returns {Object} An object containing the x and y coordinates of the
   * player avatar. 
   * @memberof Player
   */
  getPosition() {
    return {
      x: this.x,
      y: this.y
    };
  }

  /**
   * @description Calculate a random number between 0 and the specified
   * maximum integer. This function was copied from the example at
   * [MDN Javascript Math builtin object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random)
   * @param {Number} max
   * @returns {Number} A number in the range 0-`max`
   */
  getRandomInt(max) {
    return Math.floor(Math.random() * Math.floor(max));
  }

  /**
   * @description Convert a y coordinate value in the canvas to a row number in
   * the game grid.
   * @param {Number} y Value of the y coordinate in the canvas
   * @returns {Number} Row number within the game grid
   * @memberof Player
   */
  convertToRowNo(y) {
    return Math.ceil(this.y / CELL_HEIGHT);
  }

  /**
   * @description Update the player's position. This method is required by the
   * game engine.
   * @param {Number} dt a time delta between ticks of the game clock
   * @memberof Player
   */
  update() {
    // Updates to the player position are made in handleInput(), which updates
    // the x and y coordinates of the player avatar.
  }

  /**
   * @description Draw the player on the screen. This method is required by the
   * game engine.
   * @memberof Player
   */
  render() {
    ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
  }

  /**
   * @description Process a request to move the player avatar one cell on the
   * game board. Attempts to move past the left, right, top, or bottom
   * boundaries are ignored. The x and y coordinates of the players avatar
   * are calculated here rather than in the update() function to eliminate the
   * dependency on a saved direction state in update().
   * @param {String} keyCode Direction to move the player avatar. Acceptable
   * values are `left`, `right`, `up`, or `down`.
   * @memberof Player
   */
  handleInput(keyCode) {
    switch (keyCode) {
      case 'left':
        this.x = this.x <= MIN_ROW_BOUNDARY ? this.x : this.x - CELL_WIDTH;
        break;
      case 'up':
        this.y = this.y <= MIN_COL_BOUNDARY ? this.y : this.y - CELL_HEIGHT;
        break;
      case 'right':
        this.x = this.x >= MAX_ROW_BOUNDARY ? this.x : this.x + CELL_WIDTH;
        break;
      case 'down':
        this.y = this.y >= MAX_COL_BOUNDARY ? this.y : this.y + CELL_HEIGHT;
        break;
      default:
        throw new Error(`Invalid key code encountered - ${keyCode}`);
    }
  }

  /**
   * @description Check to see if the player has won the current game
   * @returns {Boolean} `true` if the game has been won
   * @memberof Player
   */
  hasWon() {
    if (this.convertToRowNo(this.y) === WIN_ROW) {
      this.resetPosition();
      return true;
    }
    return false;
  }
}

js/Game.js

/**
 * @description Game defines the player sprite and controls its movement
 * game board
 * @class Player
 */
class Game {
  /**
   * @description Creates an instance of Game and creates and initializes
   * object variables. The modal dialog code included in this class was
   * adapted from the Udacity Front End Developer Nanodegree [lesson on 
   * ARIA](https://classroom.udacity.com/nanodegrees/nd001/parts/c02fda3b-67bf-48d6-a64f-c6960e2d4d79/modules/d91b4314-da9f-45ea-902e-0b1fb5a06c34/lessons/8311490720/concepts/85569184480923)
   * @memberof Game
   */
  constructor() {
    // Game metrics used to maintain the player's expertise and performance
    // at playing the game.

    // Initialize the game score and number 
    this.noGamesWon = 0;
    this.noGamesPlayed = -1;
    this.awardedGems = 0;

    // Image URLs for the gems that can be awarded during game play
    this.gemUrls = [
      'images/Gem Blue.png',
      'images/Gem Green.png',
      'images/Gem Orange.png',
    ];

    // Define the gems to be awarded at specific win percentage thresholds
    this.gemAwards = [
      { threshold: 0, noGems: 0 },
      { threshold: 9, noGems: 1 },
      { threshold: 32, noGems: 2 },
      { threshold: 65, noGems: 3 }
    ];

    // Add the gems to the page, but hide them until earned
    for (let i = 0; i < MAX_GEMS; i += 1) {
      document.querySelector(`#gem-${i}`).setAttribute('src', this.gemUrls[i]);
      document.querySelector(`#gem-${i}`).style.display = 'none';
    }

    // Will hold previously focused element
    this.focusedElementBeforeModal = null;

    // Find the modal and its overlay
    this.modal = document.querySelector('.modal');
    this.modalOverlay = document.querySelector('.modal-overlay');
    this.closeModal = null;
  }

  /**
   * @description Retrieve the current game score - games won and games played
   * @returns {Object} An object containing the number of games won and played.
   * @memberof Game
   */
  getScore() {
    return {
      gamesWon: this.noGamesWon,
      gamesPlayed: this.noGamesPlayed
    };
  }

  /**
   * @description Increment the number of games won
   * @returns {Number} Number of games won
   * @memberof Game
   */
  incrementWins() {
    this.noGamesWon += 1;
    return this.noGamesWon;
  }

  /**
   * @description Update the number of awarded gems based on the players 
   * win percentage.
   * @memberof Game
   */
  updateGems() {
    // Calculate the number of gems to be awarded based on the players
    // win percentage
    const winPercentage = (this.noGamesWon === 0 || this.noGamesPlayed === 0)
      ? 0
      : (this.noGamesWon / this.noGamesPlayed).toFixed(2) * 100;
    this.awardedGems = this.gemAwards.reduce((gemCount, percentileRange) => {
      if (winPercentage >= percentileRange.threshold) {
        gemCount = percentileRange.noGems;
      }
      return gemCount;
    }, 0);

    // Add gems that have been awarded to the page
    for (let i = 0; i < MAX_GEMS; i += 1) {
      if (i < this.awardedGems) {
        document.querySelector(`#gem-${i}`).style.display = 'block';
      } else {
        document.querySelector(`#gem-${i}`).style.display = 'none';
      }
    }
  }

  /**
   * @description Increment the number of games played
   * @returns {Number} Number of games played
   * @memberof Game
   */
  incrementPlays() {
    this.noGamesPlayed += 1;
    return this.noGamesPlayed;
  }
  
  /**
   * @description Open the modal dialog
   * @param {String} messageText Text to be displayed in the dialogs message area
   * @memberof Game
   */
  openModal(messageText) {
    // Save current focus
    this.focusedElementBeforeModal = document.activeElement;

    // Customize the dialog message
    document.querySelector('#dialog-message').innerHTML = messageText;

    // Listen for and trap keyboard events. Bind the trap function to the
    // context of 'this' object rather than that of the event
    this.trapSpecialKeys = this.trapTheSpecialKeys.bind(this);
    this.modal.addEventListener('keydown', this.trapSpecialKeys);

    // Listen for indicators to close the modal. Bind the trap function to the
    // context of 'this' object rather than that of the button
    this.closeModal = this.closeTheModal.bind(this);
    this.modalOverlay.addEventListener('click', this.closeModal);
    const continueButton = this.modal.querySelector('#continue-button');
    continueButton.addEventListener('click', this.closeModal);

    // Find all focusable children
    const focusableElementsString = 'a[href], area[href], input:not([disabled]), select:not([disabled]), textarea:not([disabled]), button:not([disabled]), iframe, object, embed, [tabindex="0"], [contenteditable]';
    let focusableElements = this.modal.querySelectorAll(focusableElementsString);
    focusableElements = Array.prototype.slice.call(focusableElements);
    const firstTabStop = focusableElements[0];
    const lastTabStop = focusableElements[focusableElements.length - 1];
    // Add the focusable tab stops to the document so they can be referenced
    // by the key handler
    document.firstTabStop = firstTabStop;
    document.lastTabStop = lastTabStop;

    // Show the modal and overlay
    this.modal.style.display = 'block';
    this.modalOverlay.style.display = 'block';

    // Focus first child
    firstTabStop.focus();

    document.querySelector('.wrapper').setAttribute('aria-hidden', true);
  }

  /**
   * @description Trap the tab and escape key
   * @param {object} event Instance of an event object
   * @memberof Game
   */
  trapTheSpecialKeys(event) {
    // Check for TAB key press
    if (event.keyCode === 9) {

      // SHIFT + TAB
      if (event.shiftKey) {
        if (document.activeElement === document.firstTabStop) {
          event.preventDefault();
          document.lastTabStop.focus();
        }

      // TAB
      } else {
        if (document.activeElement === document.lastTabStop) {
          event.preventDefault();
          document.firstTabStop.focus();
        }
      }
    }

    // Close the modal if an ESC key is pressed
    if (event.keyCode === 27) {
      this.closeTheModal();
    }
  }

  /**
   * @description Close the open modal dialog
   * @memberof Game
   */
  closeTheModal(event) {
    const continueButton = event.target.closest('#continue-button');
    if (continueButton !== null) {
      this.updateGems();

      // Hide the modal and overlay
      this.modal.style.display = 'none';
      this.modalOverlay.style.display = 'none';

      // Set focus back to element that had it before the modal was opened
      this.focusedElementBeforeModal.focus();

      document.querySelector('.wrapper').removeAttribute('aria-hidden');
    }
  }
}